tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -2966,84 +2967,125 @@ llvm::Optional<Value> convertResizeOp(PatternRewriter& rewriter, Operation* op,
   size_t output_height = output_shape[1];
   size_t output_width = output_shape[2];
 
-  double fp_stride_y =
-      static_cast<double>(input_height) / static_cast<double>(output_height);
-  double fp_stride_x =
-      static_cast<double>(input_width) / static_cast<double>(output_width);
-  if (align_corners && output_height > 1) {
-    fp_stride_y = static_cast<double>(input_height - 1) /
-                  static_cast<double>(output_height - 1);
-  }
-  if (align_corners && output_width > 1) {
-    fp_stride_x = static_cast<double>(input_width - 1) /
-                  static_cast<double>(output_width - 1);
-  }
-
-  double fp_offset_y, fp_offset_x;
-  if (half_pixel_centers) {
-    fp_offset_y = fp_stride_y * 0.5f - 0.5f;
-    fp_offset_x = fp_stride_x * 0.5f - 0.5f;
-  } else {
-    fp_offset_y = 0.0f;
-    fp_offset_x = 0.0f;
-  }
+  auto isPowerOfTwoUpscale = [](size_t a, size_t b) {
+    auto isPowerOfTwo = [](size_t x) { return x && (!(x & (x - 1))); };
+    return (a % b == 0) && isPowerOfTwo(a / b);
+  };
+
+  // a. For power of two upscale [OH, OW] = (1 << k) * [IH, IW],
+  // sampling range approximately (-0.5, -0.5) to (IH - 0.5, IW - 0.5), set:
+  // scale_y_n = 2 << k, scale_y_d = 2, offset_y = -(1 << k) + 1, border_y = (1
+  // << k)-1 scale_x_n = 2 << k, scale_x_d = 2, offset_x = -(1 << k) + 1,
+  // border_x = (1 << k)-1
+
+  // b. For power of two upscale [OH - 1 ,OW - 1] = (1 << k) * [IH - 1, IW - 1],
+  // sampling between (0,0) and (IH - 1,IW - 1), set: scale_y_n = (1 << k),
+  // scale_y_d = 1, offset_y = 0, border_y = 0 scale_x_n = (1 << k), scale_x_d =
+  // 1, offset_x = 0, border_x = 0
+
+  // c. For approximate uniform input sampling between (0, 0) and (IH - 1, IW -
+  // 1) set scale_y_n/scale_y_d = (OH - 1)/(IH - 1) as integer ratios
+  // scale_x_n/scale_x_d = (OW - 1)/(IW - 1) as integer ratios
+  // offset_x = 0, offset_y = 0, border_x = 0, border_y = 0
+
+  int scale_y_n, scale_y_d, scale_x_n, scale_x_d;
+  int offset_y = 0, offset_x = 0;
+  int border_y = 0, border_x = 0;
+  bool uniform_sampling = true;
+
+  if (isPowerOfTwoUpscale(output_height, input_height) &&
+      isPowerOfTwoUpscale(output_width, input_width) &&
+      (output_height / input_height == output_width / input_width)) {
+    int n = output_height / input_height;
+    int k = std::log2(n & -n);
```

Comment:
Could you add a comment for this?

Reply:
Thanks for spotting this. The calculation here is to evaluate the exponent. since the value at this point is already power-of-two, there is only one set bit in the integer value. I think it can be simplified to `log(n)`. I've added a comment in the next patch. Feel free to let me know if this is enough.